### ISSUE TITLE
Confidence / Error bars for shap values

### ISSUE THREAD

**Issue by nayan1247** [shap/shap#441]
Hi,

What would be a best way to compute error bars / confidence intervals around the shap values (with focus on main effect values). It might not be possible to estimate it directly, but was wondering if there is an alternate way where a proxy metric can be computed. 

Background - Along with the shap values for the observation, in the main effect plot want to bring in the error bars as well. Knowing that shap is just as accurate, the the model it might be useful in someway to use the model error for observations in combination with shap feature values to get the proxy metric for error bars/confidence intervals.I am using xgboost with treeshap explainer.

Any thoughts on the above ?

Thanks in Advance !

**Comment by slundberg** [comment]
I use bootstrap resampling to estimate confidence intervals. Train many models on bootstraps of the train set then see the variability of those model's predictions (or SHAP values).

**Comment by nayan1247** [comment]
Thanks. Will try this out.

**Comment by cjpeck** [comment]
One thing that I have been playing around with is to use quantile regression in LightGBM to get 3 models that predict the 5%, 50%, and 95% quantile of the predictions e.g.
```python
model_lower = lgb.LGBMRegressor(objective='quantile', alpha=0.05)
model_median = lgb.LGBMRegressor(objective='quantile', alpha=0.5)
model_upper = lgb.LGBMRegressor(objective='quantile', alpha=0.95)
model_lower.fit(X,y)
...
```

And then get the shap values for each of these 3 models to generate the error bars i.e.
```python
shap_dict = {}
for label, model in [("upper", model_upper), ("median", model_median), ("lower", model_lower)]:
    explainer = shap.TreeExplainer(model)
    shap_dict[label] = explainer.shap_values(X)
```

I'm still not convinced whether or not these could really be considered 'confidence intervals' of the SHAP values as you would get w/ bootstrapping. Planning on doing a simulation of this next week to investigate. @slundberg does this seem like a reasonable approach to you?

**Comment by slundberg** [comment]
It seems reasonable to explain (and then compare) a 5% vs a 95% quantile model. But that just tells you how things change at opposite tails of the distribution. It does not tell you how confident you are in your explanation. So while interesting, I would not consider it a confidence interval alternative to bootstrapping.

**Comment by andrejastojic** [comment]
Hi Scott,

I'd like to calculate errors/stability for SHAP values, for example applying bootstrap on dependence_plot function. 

How can I track each data point from my data set (first, second, etc) to see the difference in its attribution values - as I can see, the ordering of the function output (Predictor 1 value, SHAP vaule and Predictor 2 value) changes in each bootstrap iteration? Can I achieve same ordering so I could get all outputs arranged from the first to the last value of Predictor 1 according to my data set (and corresponding SHAP and Predictor 2 values)?

Also, I don't understand why "we randomize the ordering so plotting overlaps are not related to data ordering" - dependence_plot function. Can you explain in more details, please?

Thanks in advance!
Andreja

**Comment by mosscoder** [comment]
> I use bootstrap resampling to estimate confidence intervals. Train many models on bootstraps of the train set then see the variability of those model's predictions (or SHAP values).

In the context of global variable importance for main effects (mean abs SHAP values), would it make sense to simply bootstrap the predictions from a single model to derive CI's?

**Comment by angelotc** [comment]
Interested in this. Interested to know if SHAP value error bars can be used as an uncertainty quantification technique.

**Comment by johege** [comment]
I'm also interested in using SHAP value error bars for uncertainty quantification.
One question that I have about the bootstrapping approach:
When I train two models on two different bootstrapped training datasets, the expected model outputs (the "base_value") will be different. Since the feature's SHAP-values "explain" the difference between a predicted value and the expected model output, the SHAP-values for the two models don't really "explain" the same thing. Should I make some normalization to the SHAP-values of the different models, e.g., by dividing by the difference between prediction and base_value?
